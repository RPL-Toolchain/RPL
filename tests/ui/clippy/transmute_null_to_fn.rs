//@revisions: inline normal
//@[inline]compile-flags: -Z mir-opt-level=0 -Z inline-mir=true
//@[normal]compile-flags: -Z mir-opt-level=0 -Z inline-mir=false
#![allow(dead_code)]

// Easy to lint because these only span one line.
fn one_liners() {
    unsafe {
        let _: fn() = std::mem::transmute(0 as *const ());
        //~^ transmute_null_to_fn

        let _: fn() = std::mem::transmute(std::ptr::null::<()>());
        //~^ transmute_null_to_fn
    }
}

pub const ZPTR: *const usize = 0 as *const _;
pub const NOT_ZPTR: *const usize = 1 as *const _;

fn transmute_const() {
    unsafe {
        // Should raise a lint.
        let _: fn() = std::mem::transmute(ZPTR);
        //~^ transmute_null_to_fn

        // Should NOT raise a lint.
        let _: fn() = std::mem::transmute(NOT_ZPTR);
    }
}

// #[rpl::dump_mir(dump_cfg, dump_ddg)]
fn issue_11485() {
    unsafe {
        let _: fn() = std::mem::transmute(0 as *const u8 as *const ());
        //~^ transmute_null_to_fn

        let _: fn() = std::mem::transmute(std::ptr::null::<()>() as *const u8);
        //~^ transmute_null_to_fn

        let _: fn() = std::mem::transmute(ZPTR as *const u8);
        //~^ transmute_null_to_fn
    }
}

fn main() {
    one_liners();
    //~[inline]^ transmute_null_to_fn
    //~[inline]| transmute_null_to_fn
    transmute_const();
    //~[inline]^ transmute_null_to_fn
}
