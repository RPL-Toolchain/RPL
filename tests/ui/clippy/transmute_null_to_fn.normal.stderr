error: transmuting a known null pointer into a function pointer
  --> tests/ui/clippy/transmute_null_to_fn.rs:9:23
   |
LL |         let _: fn() = std::mem::transmute(0 as *const ());
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try wrapping your function pointer type in `Option<T>` instead, and using `None` as a null pointer value
   = note: `0 as *const ()` is known to be a null pointer, so transmuting it into a function pointer is undefined behavior
   = note: `-D rpl::transmute-null-to-fn` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::transmute_null_to_fn)]`

error: transmuting a known null pointer into a function pointer
  --> tests/ui/clippy/transmute_null_to_fn.rs:12:23
   |
LL |         let _: fn() = std::mem::transmute(std::ptr::null::<()>());
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try wrapping your function pointer type in `Option<T>` instead, and using `None` as a null pointer value
   = note: `std::ptr::null::<()>()` is known to be a null pointer, so transmuting it into a function pointer is undefined behavior

error: transmuting a known null pointer into a function pointer
  --> tests/ui/clippy/transmute_null_to_fn.rs:23:23
   |
LL |         let _: fn() = std::mem::transmute(ZPTR);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try wrapping your function pointer type in `Option<T>` instead, and using `None` as a null pointer value
   = note: `ZPTR` is known to be a null pointer, so transmuting it into a function pointer is undefined behavior

error: transmuting a known null pointer into a function pointer
  --> tests/ui/clippy/transmute_null_to_fn.rs:40:23
   |
LL |         let _: fn() = std::mem::transmute(ZPTR as *const u8);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try wrapping your function pointer type in `Option<T>` instead, and using `None` as a null pointer value
   = note: `ZPTR as *const u8` is known to be a null pointer, so transmuting it into a function pointer is undefined behavior

error: transmuting a known null pointer into a function pointer
  --> tests/ui/clippy/transmute_null_to_fn.rs:34:23
   |
LL |         let _: fn() = std::mem::transmute(0 as *const u8 as *const ());
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try wrapping your function pointer type in `Option<T>` instead, and using `None` as a null pointer value
   = note: `0 as *const u8 as *const ()` is known to be a null pointer, so transmuting it into a function pointer is undefined behavior

error: transmuting a known null pointer into a function pointer
  --> tests/ui/clippy/transmute_null_to_fn.rs:37:23
   |
LL |         let _: fn() = std::mem::transmute(std::ptr::null::<()>() as *const u8);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try wrapping your function pointer type in `Option<T>` instead, and using `None` as a null pointer value
   = note: `std::ptr::null::<()>()` is known to be a null pointer, so transmuting it into a function pointer is undefined behavior

error: aborting due to 6 previous errors

